src/uct/ib/rc/accel/rc_mlx5_ep.c

```diff
@@ -253,6 +254,122 @@ ucs_status_t uct_rc_mlx5_ep_am_short(uct_ep_h tl_ep, uint8_t id, uint64_t hdr,
     return UCS_OK;
 }
 
+#if HAVE_IBV_EXP_DM
+/* DM memory should be written by 8 bytes (int64) to eliminate
+ * processor cache issues. To make this used uct_rc_mlx5_dm_copy_data_t
+ * datatype where first hdr_len bytes are filled by message header
+ * and tail is filled by head of message. */
+static void UCS_F_ALWAYS_INLINE
+uct_rc_mlx5_ep_copy_to_dm(uct_rc_mlx5_dm_copy_data_t *cache, size_t hdr_len,
+                          const void *payload, size_t length, void *dm)
+{
+    size_t len          = sizeof(*cache) - hdr_len;
+    const uint64_t *src = payload + len;
+    uint64_t *dst       = dm;
+
+    ucs_assert(sizeof(*cache) >= hdr_len);
+
+    /* copy head of payload to tail of cache */
+    memcpy(cache->in + hdr_len, payload, ucs_min(len, length));
+
+    UCS_STATIC_ASSERT(sizeof(*cache) == sizeof(cache->out));
+    UCS_STATIC_ASSERT(sizeof(cache->in) == sizeof(cache->out));
+
+    /* atomically by 8 bytes copy data to DM */
+    *(dst++) = cache->out[0];
+    *(dst++) = cache->out[1];
+
+    for (; len < length; len += sizeof(*dst)) {
+        *(dst++) = *(src++);
+    }
+}
+
+static ucs_status_t UCS_F_ALWAYS_INLINE
+uct_rc_mlx5_ep_short_dm(uct_rc_mlx5_ep_t *ep, uct_rc_mlx5_dm_copy_data_t *cache,
+                        size_t hdr_len, const void *payload, unsigned length, int stat)
+{
+    uct_rc_mlx5_iface_t *iface = ucs_derived_of(ep->super.super.super.iface, uct_rc_mlx5_iface_t);
+    uct_rc_iface_t *rc_iface   = &iface->super;
+    uct_rc_iface_send_desc_t *desc;
+
+    desc = ucs_mpool_get_inline(&iface->mlx5_common.dm.dm->mp);
+    if (ucs_unlikely((desc) == NULL)) {
+        /* in case if no resources available - fallback to bcopy */
+        UCT_RC_IFACE_GET_TX_DESC(rc_iface, &rc_iface->tx.mp, desc);
+        desc->super.handler = (uct_rc_send_handler_t)ucs_mpool_put;
+
+        memcpy(desc + 1, cache->out, hdr_len);
+        memcpy(UCS_PTR_BYTE_OFFSET(desc + 1, hdr_len), payload, length);
+
+        uct_rc_mlx5_txqp_bcopy_post(rc_iface, &ep->super.txqp, &ep->tx.wq,
+                                    MLX5_OPCODE_SEND, hdr_len + length,
+                                    0, 0, MLX5_WQE_CTRL_SOLICITED, 0, desc, desc + 1);
+        if (stat) {
+            UCT_TL_EP_STAT_OP(&ep->super.super, AM, BCOPY, length);
+        }
+    } else {
+        ucs_assert(desc->super.buffer != NULL);
+
+        uct_rc_mlx5_ep_copy_to_dm(cache, hdr_len, payload,
+                                  length, desc->super.buffer);
+
+        uct_rc_mlx5_txqp_bcopy_post(rc_iface, &ep->super.txqp, &ep->tx.wq,
```

Comment:
to avoid some code duplication, can set "send_flags" and "buffer_pointer" inside if/else, and then call uct_rc_mlx5_txqp_bcopy_post()+stat_op outside the if/else.